wrappers.ts

```diff
@@ -46,28 +47,281 @@ const withLoggedArguments =
 
 */
 
-// TODO middleware should have a required type
-type wrap1 = <T>(wf: T) => T
-type wrap2 = <T>(mw1: any, wf: T) => T
-type wrap3 = <T>(mw1: any, mw2: any, wf: T) => T
-type wrap4 = <T>(mw1: any, mw2: any, mw3: any, wf: T) => T
-type wrap5 = <T>(mw1: any, mw2: any, mw3: any, mw4: any, wf: T) => T
-type wrap6 = <T>(mw1: any, mw2: any, mw3: any, mw4: any, mw5: any, wf: T) => T
-type wrap7 = <T>(mw1: any, mw2: any, mw3: any, mw4: any, mw5: any, mw6: any, wf: T) => T
+export type Middleware<T, Dep = {}> = (
+  next: (req: Req & Dep & T, res: Res) => any
+) => (req: Req & Dep & T, res: Res) => any
 
-type Wrappers = wrap1 & wrap2 & wrap3 & wrap4 & wrap5 & wrap6 & wrap7
+// Safer Middleware requires the use of extendRequest to ensure that the
+// new context (T) was actually added to the request. It's kind of annoying
+// to use in practice, so we don't use it for our Wrappers (yet)
+export type SaferMiddleware<T, Dep = {}> = (
+  next: (req: Req & Dep & T, res: Res) => any
+) => (req: Req & Dep, res: Res) => any
 
-export const wrappers: Wrappers = (...wrappersArgs) => {
+export const extendRequest = <T extends Req, K>(req: T, merge: K): T & K => {
+  for (const [key, v] of Object.entries(merge)) {
+    ;(req as any)[key] = v
+  }
+  return req as any
+}
+
+type Wrappers1 = <Mw1RequestContext, Mw1Dep>(
+  mw1: Middleware<Mw1RequestContext, Mw1Dep>,
+  endpoint: (req: Mw1RequestContext, res: any) => any
```

Comment:
can we type `res` as `NextApiResponse` instead of `any`?

Reply:
yep!